core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java

```diff
@@ -185,7 +186,15 @@ private void performOnReplica(final ShardRouting shard, final ReplicaRequest rep
             @Override
             public void onResponse(ReplicaResponse response) {
                 successfulShards.incrementAndGet();
-                primary.updateLocalCheckpointForShard(response.allocationId(), response.localCheckpoint());
+                try {
+                    primary.updateLocalCheckpointForShard(response.allocationId(), response.localCheckpoint());
+                } catch (final AlreadyClosedException e) {
+                    // okay, the index was deleted or this shard was never activated after a relocation; fallthrough and finish normally
+                } catch (final Exception e) {
+                    final String message = String.format(Locale.ROOT, "primary failed updating local checkpoint for replica %s", shard);
+                    primary.failShard(message, e);
+                    finishAsFailed(new RetryOnPrimaryException(primary.routingEntry().shardId(), message, e));
```

Comment:
I don't think we should do this. The failure is not related to this operation. IMO we should keep processing the operation as normal. This would be identical to the primary failing just after we acked the write (for whatever reason)